module/zfs/vfs_ratelimit.c

```diff
@@ -0,0 +1,688 @@
+ratelimit_parent(dsl_dir_t *dd, int types)
+{
+	ASSERT(RRM_READ_HELD(&dd->dd_pool->dp_spa->spa_ratelimit_lock));
+
+	for (;;) {
+		if (dd->dd_parent == NULL) {
+			return (NULL);
+		}
+		dd = dd->dd_parent->dd_ratelimit_root;
+		if (dd == NULL) {
+			return (NULL);
+		}
+		if (dd->dd_ratelimit != NULL) {
+			int mytypes;
+
+			mytypes = ratelimit_types(dd->dd_ratelimit->rl_limits);
+			if ((mytypes & types) != 0) {
+				/*
+				 * This dataset has at last one limit we are
+				 * interested in.
+				 */
+				return (dd);
+			}
+		}
+	}
+}
+
+/*
+ * Account for our request across all the types configured in this ratelimit
+ * structure.
+ * Return a timeslot we should wait for or now if we can execute the request
+ * without waiting (we are within limits).
+ */
+static hrtime_t
+ratelimit_account(struct vfs_ratelimit *rl, hrtime_t now,
+    const uint64_t *counts)
+{
+	hrtime_t timeslot;
+	int type;
+
+	timeslot = now;
+
+	mutex_enter(&rl->rl_lock);
+
+	for (type = ZFS_RATELIMIT_FIRST; type <= ZFS_RATELIMIT_LAST; type++) {
+		uint64_t count;
+
+		if (rl->rl_limits[type] == 0) {
+			/* This type has no limit configured on this dataset. */
+			continue;
+		}
+		count = counts[type];
+		if (count == 0) {
+			/* Not interested in this type. */
+			continue;
+		}
+
+		if (rl->rl_timeslot[type] < now) {
+			rl->rl_reminder[type] = 0;
+			rl->rl_timeslot[type] = now;
+		} else {
+			count += rl->rl_reminder[type];
+		}
+
+		rl->rl_timeslot[type] += count / rl->rl_limits[type];
+		rl->rl_reminder[type] = count % rl->rl_limits[type];;
+
+		if (timeslot < rl->rl_timeslot[type]) {
+			timeslot = rl->rl_timeslot[type];
+		}
+	}
+
+	mutex_exit(&rl->rl_lock);
+
+	return (timeslot);
+}
+
+static hrtime_t
+ratelimit_account_all(objset_t *os, const uint64_t *counts)
+{
+	dsl_dir_t *dd;
+	hrtime_t now, timeslot;
+	int types;
+
+	ASSERT(RRM_READ_HELD(&os->os_spa->spa_ratelimit_lock));
+
+	types = ratelimit_types(counts);
+	now = timeslot = gettimeslot();
+
+	for (dd = ratelimit_first(os, types); dd != NULL;
+	    dd = ratelimit_parent(dd, types)) {
+		hrtime_t ts;
+
+		ts = ratelimit_account(dd->dd_ratelimit, now, counts);
+		if (ts > timeslot) {
+			timeslot = ts;
+		}
+	}
+
+	return (timeslot);
+}
+
+static int
+ratelimit_sleep(hrtime_t timeslot)
+{
+	hrtime_t now;
+	int error = 0;
+
+	now = gettimeslot();
+
+	if (timeslot > now) {
+		/*
+		 * Too much traffic, slow it down.
+		 */
+#ifdef _KERNEL
+		if (delay_sig((hz / RATELIMIT_RESOLUTION) * (timeslot - now))) {
+			error = EINTR;
+		}
+#else
+		delay((hz / RATELIMIT_RESOLUTION) * (timeslot - now));
+#endif
+	}
+
+	return (error);
+}
+
+static int
+vfs_ratelimit_sleep(objset_t *os, const uint64_t *counts)
+{
+	hrtime_t timeslot;
+
+	/*
+	 * Prevents configuration changes when we have requests in-flight.
+	 */
+	rrm_enter_read(&os->os_spa->spa_ratelimit_lock, FTAG);
+
+	timeslot = ratelimit_account_all(os, counts);
+
+	rrm_exit(&os->os_spa->spa_ratelimit_lock, FTAG);
+
+	return (ratelimit_sleep(timeslot));
```

Comment:
I think it would be nice if in case of abort we undo accounting done, otherwise following requests will still wait for whatever insane time we could have been given.